File: c/problem_0025.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>
#include "./lib/mathlib.h"

/**
 * @see https://projecteuler.net:
 *
 * 1000-digit Fibonacci number
 * Problem 25
 *
 * The Fibonacci sequence is defined by the recurrence relation:
 *
 *     Fn = Fn−1 + Fn−2, where F1 = 1 and F2 = 1.
 *
 * Hence the first 12 terms will be:
 *
 *     F1 = 1
 *     F2 = 1
 *     F3 = 2
 *     F4 = 3
 *     F5 = 5
 *     F6 = 8
 *     F7 = 13
 *     F8 = 21
 *     F9 = 34
 *     F10 = 55
 *     F11 = 89
 *     F12 = 144
 *
 * The 12th term, F12, is the first term to contain three digits.
 *
 * What is the index of the first term in the Fibonacci sequence to contain 1000 digits?
 */

int main(void)
{
    bignum_int *results[3];
    int number_index = 2;
    results[0] = bignum_int_create(1000, 1);
    results[1] = bignum_int_create(1000, 1);
    // results[2] = bignum_int_sum(results[0], results[1], 1000);

    while (results[(number_index - 1) % 3]->length < 1000)
    {
        bignum_int *tmp = bignum_int_sum(
            results[(number_index - 1) % 3],
            results[(number_index - 2) % 3], 1000);
        results[number_index % 3] = tmp;
        number_index++;
    }
    printf("First Fibonacci Number with 1000 digits: #%d: %s\n", number_index, bignum_int_str(results[(number_index - 1) % 3]));
}

File: haskell/src/Problem0025.hs
module Problem0025 (solution) where

{--------------------------------------------------------------------
<p>The Fibonacci sequence is defined by the recurrence relation:</p>
<blockquote>$F_n = F_{n - 1} + F_{n - 2}$, where $F_1 = 1$ and $F_2 = 1$.</blockquote>
<p>Hence the first $12$ terms will be:</p>
\begin{align}
F_1 &amp;= 1\\
F_2 &amp;= 1\\
F_3 &amp;= 2\\
F_4 &amp;= 3\\
F_5 &amp;= 5\\
F_6 &amp;= 8\\
F_7 &amp;= 13\\
F_8 &amp;= 21\\
F_9 &amp;= 34\\
F_{10} &amp;= 55\\
F_{11} &amp;= 89\\
F_{12} &amp;= 144
\end{align}
<p>The $12$th term, $F_{12}$, is the first term to contain three digits.</p>
<p>What is the index of the first term in the Fibonacci sequence to contain $1000$ digits?</p>
---------------------------------------------------------------------}

import EulerLib qualified as E

solution :: E.ProblemSolution
solution =
  E.ProblemSolution
    { E.psNr = 25,
      E.psTitle = "1000-digit Fibonacci Number",
      E.psSolve = \_ -> show endResult,
      E.psSolution = "",
      E.psReadInput = pure ""
    }

minDigits :: Int
minDigits = 1000

fiboList :: [Integer]
fiboList = E.fiboTo (\x -> length (show x) == minDigits)

endResult :: Int
endResult = length fiboList + 1